notes.md

```diff
@@ -0,0 +1,264 @@
+in the code.
+
+I have tested as much as possible without having postgres to unit test over.
+There will still be a lot of bugs and probably code which could be refactored
+and improved.
+The examples below are actually working examples at least.
+
+
+# Documentation/Notes via Examples
+Here we show how a user can use django-pubsub to set up and notify listeners
+to process functions async. We'll make use of the models and relationships
+defined in the tests app.
+
+We focus on two use cases:
+- The first is where the user defines a listener with custom keyword arguments and notifies
+  that listener manually.
+- The second is where the user defines a listener which is notified whenever some django-pgtrigger is executed.
+
+Fully working versions of these examples can be found in the `tests` app.
+
+## Custom Listen and Manual Notify
+
+In this example, we wish to maintain a cache of how frequently
+Post objects are read per day. We want to defer this caching logic
+to another process so as not to hurt read performance times for posts.
+The user achieves this by first implementing a listener for the caching:
+
+Note: This example is for illustrative purposes only. There are of course
+many other ways to achieve this.
+
+
+### Declaring a Listener
+First, define the listener in their app's `tests.listeners.py` file:
+
+```python
+post_reads_per_date_cache = defaultdict(dict)  # dummy cache for illustrative purposes only
+
+@listen('post_reads_per_date')
+def post_reads_per_date(
+        *,
+        post_id: int,
+        date: datetime.date,
+):
+    print(f'Processing post_reads_per_date with args {post_id}, {date}')
+    print(f'Cache before: {post_reads_per_date_cache}')
+    current_count = post_reads_per_date_cache[date].get(post_id, 0)
+    post_reads_per_date_cache[date][post_id] = current_count + 1
+    print(f'Cache after: {post_reads_per_date_cache}')
+```
+
+The user must also ensure that this `listeners` module is imported into
+that app's config class (similar to signals):
+```python
+class TestsConfig(AppConfig):
+    name = 'tests'
+
+    def ready(self):
+        import tests.listeners
+```
+
+#### Registering listeners
+When the project is loaded, the `listen` decorator is called. This
+executes this piece of logic:
+```python
+CustomPayloadChannel.register(callback, channel_name)
+```
+This creates an instance of CustomPayloadChannel and adds it to a global
+registry of "channels".
+
+##### Channels
+As in the postgres listen/notify protocol, messages are sent and received
+via named channels. The objects defined in channels.py are python abstractions
+of these channels. Instances of these classes have the following responsibilities:
+
+- Define an appropriate channel name if one has not been supplied.
+    - Undecided on how to best generate the name - see comments in _create_name.
+    - This is why in the examples I just define the names manually.
```

Comment:
Another possibility than naming channels is to have users define a `Channel` class first, for example:

```
channel = MyChannel()
channel.notify(...)


# This is what a listener looks like
@listener(MyChannel)
def listen(...):
    pass
```

I generally avoid libraries that have the potential to introduce subtle human error, so listening to either Channel class names or channel instances is a nice way to avoid mistyping a string label.

Another potential benefit this could offer is hierarchy - I.e. the `listen` function will listen to any class AND subclass of MyChannel, for example.

Going a bit deeper into the rabbit hole - consider you want your listen/notify to have a particular payload in a channel. A channel definition would allow you to define this, for example, as a dataclass:

```
@dataclass
class MyChannel:
    post_id: int
    date: datetime.date
```

Then the `listen` decorator can validate that your functional listener at least has the right call signature.

I haven't read the rest of this yet. Just commenting as I go. dataclasses make it easy to serialize to JSON, so it might be interesting to structure it this way. I.e you avoid hardcoding channel names and you define a payload schema for the channel

Reply:
If you want syntactical sugar like django and want to avoid importing classes, you could also do:


```
@listen('myapp.channels.MyChannel')
def listener(...):
    ...
```

And have it load the channel class too.